component/http/cache_builder.go

```diff
@@ -0,0 +1,130 @@
+
+import (
+	"errors"
+	"fmt"
+	"time"
+
+	"github.com/beatlabs/patron/cache"
+	errs "github.com/beatlabs/patron/errors"
+)
+
+// RouteCacheBuilder is the builder needed to build a cache for the corresponding route
+type RouteCacheBuilder struct {
+	cache         cache.Cache
+	instant       TimeInstant
+	ttl           time.Duration
+	minAge        time.Duration
+	maxFresh      time.Duration
+	staleResponse bool
+	metrics       cacheMetrics
+	errors        []error
+}
+
+// NewRouteCacheBuilder creates a new builder for the route cache implementation.
+func NewRouteCacheBuilder(cache cache.Cache, ttl time.Duration) *RouteCacheBuilder {
+
+	var ee []error
+
+	if ttl <= 0 {
+		ee = append(ee, errors.New("time to live must be greater than `0`"))
+	}
+
+	return &RouteCacheBuilder{
+		cache: cache,
+		ttl:   ttl,
+		instant: func() int64 {
+			return time.Now().Unix()
+		},
+		metrics: NewVoidMetrics(),
+		errors:  ee,
+	}
+}
+
+// WithTimeInstant specifies a time instant function for checking expiry.
+func (cb *RouteCacheBuilder) WithTimeInstant(instant TimeInstant) *RouteCacheBuilder {
+	if instant == nil {
+		cb.errors = append(cb.errors, errors.New("time instant is nil"))
+	}
+	cb.instant = instant
+	return cb
+}
+
+// WithMinAge adds a minimum age for the cache responses.
+// This will avoid cases where a single client with high request rate and no cache control headers might effectively disable the cache
+// This means that if this parameter is missing (e.g. is equal to '0' , the cache can effectively be made obsolete in the above scenario)
+func (cb *RouteCacheBuilder) WithMinAge(minAge time.Duration) *RouteCacheBuilder {
+	if minAge <= 0 {
+		cb.errors = append(cb.errors, fmt.Errorf("min-age cannot be lower or equal to zero '0 < %v'", minAge))
+	}
+	cb.minAge = minAge
+	return cb
+}
+
+// WithMaxFresh adds a maximum age for the cache responses.
+// This will avoid cases where a single client with high request rate and no cache control headers might effectively disable the cache
+// This means that if this parameter is very high (e.g. greater than ttl , the cache can effectively be made obsolete in the above scenario)
+func (cb *RouteCacheBuilder) WithMaxFresh(maxFresh time.Duration) *RouteCacheBuilder {
+	if maxFresh > cb.ttl {
+		cb.errors = append(cb.errors, fmt.Errorf("max-fresh cannot be greater than the time to live '%v <= %v'", maxFresh, cb.ttl))
+	}
+	cb.maxFresh = maxFresh
+	return cb
+}
+
+// WithStaleResponse allows the cache to return stale responses.
+func (cb *RouteCacheBuilder) WithStaleResponse(staleResponse bool) *RouteCacheBuilder {
+	cb.staleResponse = staleResponse
+	return cb
+}
+
+// WithMetrics allows the cache to return stale responses.
+func (cb *RouteCacheBuilder) WithMetrics(metrics cacheMetrics) *RouteCacheBuilder {
+	if metrics == nil {
+		cb.errors = append(cb.errors, errors.New("metrics implementation is nil"))
+	}
+	cb.metrics = metrics
+	return cb
+}
+
+func (cb *RouteCacheBuilder) create() (*routeCache, error) {
```

Comment:
It's because the user provides the route cache builder to the route builder.
We cannot create the builder before we create the route.
`func (rb *RouteBuilder) WithRouteCachedBuilder(routeCacheBuilder *RouteCacheBuilder) *RouteBuilder {
`
Does that make sense ?

Reply:
done